carla/recourse_methods/catalog/roar/library/roar.py

```diff
@@ -0,0 +1,196 @@
+    coeff: torch.Tensor,
+    intercept: torch.Tensor,
+    delta_max: float,
+    target_class: torch.Tensor,
+) -> Tuple[np.ndarray, np.ndarray]:
+    """
+    Calculate the optimal delta perturbation using linear program
+    Adjusted from: https://github.com/AI4LIFE-GROUP/ROAR/blob/main/recourse_methods.py
+
+    Parameters
+    ----------
+    recourse: Factual to explain
+    coeff: Tensor with Coefficients of linear model
+    intercept: Tensor of Intercept of linear model
+    delta_max: Maximum perturbations of weights
+    target_class: Float Tensor representing the target class
+
+    Returns
+    -------
+    Tuple of Torch tensors with optimal perturbations of coefficients and intercept
+    """
+    W = torch.cat((coeff, intercept), 0)  # Add intercept to weights
+    recourse = torch.cat(
+        (recourse, torch.ones(1)), 0
+    )  # Add 1 to the feature vector for intercept
+
+    loss_fn = torch.nn.BCELoss()
+    W.requires_grad = True
+    f_x_new = torch.nn.Sigmoid()(torch.matmul(W, recourse))
+    w_loss = loss_fn(f_x_new, target_class)
+
+    gradient_w_loss = grad(w_loss, W)[0]
+    c = list(np.array(gradient_w_loss) * np.array([-1] * len(gradient_w_loss)))
+    bound = (-delta_max, delta_max)
+    bounds = [bound] * len(gradient_w_loss)
+
+    res = linprog(c, bounds=bounds, method="simplex")
+
+    if res.status != 0:
+        log.warning("Optimization with respect to delta failed to converge")
+
+    delta_opt = res.x
+    delta_W, delta_W0 = np.array(delta_opt[:-1]), np.array([delta_opt[-1]])
+
+    return delta_W, delta_W0
+
+
+def roar_recourse(
+    torch_model,
+    x: np.ndarray,
+    coeff: np.ndarray,
+    intercept: np.float,
+    cat_feature_indices: List[int],
+    binary_cat_features: bool = True,
+    feature_costs: Optional[List[float]] = None,
+    lr: float = 0.01,
+    lambda_param: float = 0.01,
+    delta_max: float = 0.01,
+    y_target: List[int] = [0, 1],
+    t_max_min: float = 0.5,
+    norm: int = 1,
+    loss_type: str = "BCE",
+    loss_threshold: float = 1e-3,
+) -> np.ndarray:
+    """
+    Generates counterfactual example according to ROAR method for input instance x
+
+    Parameters
+    ----------
+    torch_model: black-box-model to discover
+    x: factual to explain
+    coeff: Coefficient for x
+    intercept: Intercept for x
+    cat_feature_indices: list of positions of categorical features in x
+    binary_cat_features: If true, the encoding of x is done by drop_if_binary
+    feature_costs: List with costs per feature
+    lr: learning rate for gradient descent
+    lambda_param: weight factor for feature_cost
+    delta_max: maximum perturbations of weights
+    y_target: Tuple of class probabilities (BCE loss) or [Float] for logit score (MSE loss).
+    t_max_min: maximum time of search
+    norm: L-norm to calculate cost
+    loss_type: String for loss function (MSE or BCE)
+    loss_threshold: Float threshold for loss difference between iterations
```

Comment:
Please use the same formatting as for the other methods:

Parameters
---------------
parameter: type
[\t] description